File: sdks/php/src/Resources/Webhooks.php
<?php

declare(strict_types=1);

namespace HookSniff\Resources;

use HookSniff\Request;

/**
 * HookSniff API Resource: Webhooks
 *
 * Send, list, get, replay, and batch webhooks.
 */
class Webhooks
{
    /**
     * @param array{baseUrl: string, token: string, timeout: int, numRetries: int} $ctx
     */
    public function __construct(private readonly array $ctx) {}

    /**
     * Send a single webhook.
     *
     * @param array{endpoint_id: string, event: string, data: array<string, mixed>} $input
     * @return array<string, mixed>
     * @throws \HookSniff\ApiException
     */
    public function send(array $input, ?string $idempotencyKey = null): array
    {
        $req = new Request('POST', '/v1/webhooks');
        if ($idempotencyKey !== null) {
            $req->setHeaderParam('idempotency-key', $idempotencyKey);
        }
        $req->setBody($input);
        return $req->send($this->ctx);
    }

    /**
     * Send batch webhooks.
     *
     * @param array{endpoint_id: string, events: array<int, array{event: string, data: array<string, mixed>}>} $input
     * @return array{batch_id: string, count: int}
     * @throws \HookSniff\ApiException
     */
    public function batch(array $input, ?string $idempotencyKey = null): array
    {
        $req = new Request('POST', '/v1/webhooks/batch');
        if ($idempotencyKey !== null) {
            $req->setHeaderParam('idempotency-key', $idempotencyKey);
        }
        $req->setBody($input);
        return $req->send($this->ctx);
    }

    /**
     * List deliveries.
     *
     * @return array{data: array<int, array<string, mixed>>, has_more: bool}
     * @throws \HookSniff\ApiException
     */
    public function list(?int $limit = null, ?int $offset = null): array
    {
        $req = new Request('GET', '/v1/webhooks');
        $params = [];
        if ($limit !== null) $params['limit'] = $limit;
        if ($offset !== null) $params['offset'] = $offset;
        if (!empty($params)) $req->setQueryParams($params);
        return $req->send($this->ctx);
    }

    /**
     * Get a specific delivery.
     *
     * @return array<string, mixed>
     * @throws \HookSniff\ApiException
     */
    public function get(string $id): array
    {
        $req = new Request('GET', '/v1/webhooks/{id}');
        $req->setPathParam('id', $id);
        return $req->send($this->ctx);
    }

    /**
     * Replay a delivery.
     *
     * @return array<string, mixed>
     * @throws \HookSniff\ApiException
     */
    public function replay(string $id, ?string $idempotencyKey = null): array
    {
        $req = new Request('POST', '/v1/webhooks/{id}/replay');
        $req->setPathParam('id', $id);
        if ($idempotencyKey !== null) {
            $req->setHeaderParam('idempotency-key', $idempotencyKey);
        }
        return $req->send($this->ctx);
    }
}

File: sdks/rust/src/resources/webhooks.rs
//! HookSniff API Resource: Webhooks
//!
//! Send, list, get, replay, and batch webhooks.

use crate::request::{HookSniffRequest, HookSniffRequestContext, HttpMethod};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize, Deserialize)]
pub struct WebhookSendInput {
    pub endpoint_id: String,
    pub event: String,
    pub data: Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub headers: Option<std::collections::HashMap<String, String>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WebhookBatchInput {
    pub webhooks: Vec<WebhookSendInput>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WebhookOutput {
    pub id: String,
    pub endpoint_id: String,
    pub event: String,
    pub data: Value,
    pub status: String,
    pub created_at: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_code: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_body: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attempts: Option<u64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BatchOutput {
    pub webhook_ids: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub errors: Option<Vec<BatchError>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BatchError {
    pub index: u64,
    pub error: String,
}

pub struct Webhooks {
    ctx: HookSniffRequestContext,
}

impl Webhooks {
    pub fn new(ctx: HookSniffRequestContext) -> Self {
        Self { ctx }
    }

    /// Send a webhook
    pub fn send(&self, input: &WebhookSendInput) -> Result<WebhookOutput, Box<dyn std::error::Error>> {
        let mut req = HookSniffRequest::new(HttpMethod::Post, "/v1/webhooks");
        req.set_body(input);
        req.send(&self.ctx)
    }

    /// Send a batch of webhooks
    pub fn batch(&self, input: &WebhookBatchInput) -> Result<BatchOutput, Box<dyn std::error::Error>> {
        let mut req = HookSniffRequest::new(HttpMethod::Post, "/v1/webhooks/batch");
        req.set_body(input);
        req.send(&self.ctx)
    }

    /// List webhooks
    pub fn list(&self) -> Result<Vec<WebhookOutput>, Box<dyn std::error::Error>> {
        let req = HookSniffRequest::new(HttpMethod::Get, "/v1/webhooks");
        req.send(&self.ctx)
    }

    /// Get a webhook by ID
    pub fn get(&self, id: &str) -> Result<WebhookOutput, Box<dyn std::error::Error>> {
        let mut req = HookSniffRequest::new(HttpMethod::Get, "/v1/webhooks/{id}");
        req.set_path_param("id", id);
        req.send(&self.ctx)
    }

    /// Replay a webhook
    pub fn replay(&self, id: &str) -> Result<WebhookOutput, Box<dyn std::error::Error>> {
        let mut req = HookSniffRequest::new(HttpMethod::Post, "/v1/webhooks/{id}/replay");
        req.set_path_param("id", id);
        req.send(&self.ctx)
    }
}
